File: Structural/Composite/Composite_C#/Composite2.cs
// Composite Design Pattern - Structural Category

// Definition
// Compose objects into tree structures to represent part-whole hierarchies. 
// Composite lets clients treat individual objects and compositions of objects uniformly.

// Participants
// The classes and objects participating in this pattern are:
//
// Component (DrawingElement)
// - Declares the interface for objects in the composition.
// - Implements default behavior for the interface common to all classes, as appropriate.
// - Declares an interface for accessing and managing its child components.
// - Defines an interface (optional) for accessing a component's parent in the recursive 
//   structure, and implements it if that's appropriate.
// Leaf (PrimitiveElement)
// - Represents leaf objects in the composition. A leaf has no children.
// - Defines behavior for primitive objects in the composition.
// Composite (CompositeElement)
// - Defines behavior for components having children.
// - Stores child components.
// - Implements child-related operations in the Component interface.
// Client (CompositeApp)
// - Manipulates objects in the composition through the Component interface

// MainApp startup class for Structural Composite Design Pattern.

// http://www.dofactory.com/net/composite-design-pattern

using System;
using System.Collections.Generic;

// MainApp startup class for Real-World Composite Design Pattern.
class MainApp
{
    // Entry point into console application.
    static void Main()
    {
        // Create a tree structure 
        CompositeElement root = new CompositeElement("Picture");
        root.Add(new PrimitiveElement("Red Line"));
        root.Add(new PrimitiveElement("Blue Circle"));
        root.Add(new PrimitiveElement("Green Box"));

        // Create a branch
        CompositeElement comp1 = new CompositeElement("Two Circles");
        comp1.Add(new PrimitiveElement("Black Circle"));
        comp1.Add(new PrimitiveElement("White Circle"));
        root.Add(comp1);

        // Create a branch
        CompositeElement comp2 = new CompositeElement("Two Squares");
        comp2.Add(new PrimitiveElement("Red Square"));
        comp2.Add(new PrimitiveElement("Blue Square"));
        root.Add(comp2);

        // Add and remove a PrimitiveElement
        PrimitiveElement pe1 = new PrimitiveElement("Green Line");
        root.Add(pe1);

        // Add and remove a PrimitiveElement
        PrimitiveElement pe2 = new PrimitiveElement("Yellow Line");
        root.Add(pe2);
        root.Remove(pe2);

        // Recursively display nodes
        root.Display(1);

        // Wait for user
        Console.ReadKey();
    }
}

// The 'Component' Treenode
abstract class DrawingElement
{
    protected string name;

    // Constructor
    public DrawingElement(string name)
    {
        this.name = name;
    }

    public abstract void Add(DrawingElement d);
    public abstract void Remove(DrawingElement d);
    public abstract void Display(int indent);
}

// The 'Leaf' class
class PrimitiveElement : DrawingElement
{
    // Constructor
    public PrimitiveElement(string name) : base(name) { }

    public override void Add(DrawingElement c)
    {
        Console.WriteLine("Cannot add to a PrimitiveElement");
    }

    public override void Remove(DrawingElement c)
    {
        Console.WriteLine("Cannot remove from a PrimitiveElement");
    }

    public override void Display(int indent)
    {
        Console.WriteLine(new String('-', indent) + " " + name);
    }
}

// The 'Composite' class
class CompositeElement : DrawingElement
{
    private List<DrawingElement> elements = new List<DrawingElement>();

    // Constructor
    public CompositeElement(string name) : base(name) { }

    public override void Add(DrawingElement d)
    {
        elements.Add(d);
    }

    public override void Remove(DrawingElement d)
    {
        elements.Remove(d);
    }

    public override void Display(int indent)
    {
        Console.WriteLine(new String('-', indent) + "+ " + name);

        // Display each child element on this node
        foreach (DrawingElement d in elements)
        {
            d.Display(indent + 2);
        }
    }
}

// Output:
/*
-+ Picture
--- Red Line
--- Blue Circle
--- Green Box
---+ Two Circles
----- Black Circle
----- White Circle
---+ Two Squares
----- Red Square
----- Blue Square
--- Green Line
*/

File: Structural/Composite/Composite_C++/Composite2.cpp
// Composite design pattern - Structural Category

// Structural patterns deal with decoupling the interface and implementation of classes and objects
// Composite pattern forms a tree structure of simple and composite objects 

// http://advancedcppwithexamples.blogspot.co.il/2010/09/c-example-for-composite-design-pattern.html

#include<string>
#include<vector>
#include<iostream>

using namespace std;

//The 'Component' Treenode
class Element
{
	public:

		Element(string name) : name(name) { };
		virtual void Add(Element * d) = 0;
		virtual void Remove(Element * d) = 0;
		virtual void Display(int indent) = 0;
		virtual ~Element() {};

	protected:

		string name;
		
	private:

		Element(); // disallowed
};

// The 'Leaf' class
class PrimitiveElement : public Element
{
	public:

		PrimitiveElement(string name) : Element(name) { };

		void Add(Element * d)
		{
			cout << "Cannot add to a PrimitiveElement" << endl;
		}
	
		void Remove(Element * d)
		{
			cout << "Cannot remove from a PrimitiveElement" << endl;
		}
	
		void Display(int indent)
		{
			string newStr(indent, '-');
			cout << newStr << " " << name << endl;
		}
	
		virtual ~PrimitiveElement() { };
	
	private:

		PrimitiveElement(); // not allowed
};

// The 'Composite' class
class CompositeElement : public Element
{
	public:

		CompositeElement(string name) : Element(name) { };

		void Add(Element * d)
		{
			elements.push_back(d);
		}
	
		void Remove(Element * d)
		{
			vector<Element *>::iterator it = elements.begin();
	
			while (it != elements.end())
			{
				if (*it == d)
				{
					delete d;
					elements.erase(it);
					break;
				}

				++it;
			}
		}

		void Display(int indent)
		{
			string newStr(indent, '-');

			cout << newStr << "+ " << name << endl;

			vector<Element *>::iterator it = elements.begin();
	
			while(it != elements.end())
			{
				(*it)->Display(indent + 2);
				++it;
			}
		}
	
		virtual ~CompositeElement()
		{
			while(!elements.empty())
			{
				vector<Element *>::iterator it = elements.begin();
				delete *it;
				elements.erase(it);
			}
		}
	
	private:
	
		CompositeElement(); // not allowed

		vector<Element *> elements;

};

int main()
{
	// Create a Tree Structure
	CompositeElement * root = new CompositeElement("Paintings");
	root->Add(new PrimitiveElement("Storm"));
	root->Add(new PrimitiveElement("Seashore"));
	root->Add(new PrimitiveElement("Night in Venice"));
	root->Add(new PrimitiveElement("Ninth Wave"));

	// Create first Branch
	CompositeElement * comp1 = new CompositeElement("Geometric figures");
	comp1->Add(new PrimitiveElement("Black Circle"));
	comp1->Add(new PrimitiveElement("White Triangle"));
	comp1->Add(new PrimitiveElement("Red Square"));
	comp1->Add(new PrimitiveElement("Blue Line"));
	root->Add(comp1);

	// Create second Branch
	CompositeElement * comp2 = new CompositeElement("Animals");
	comp2->Add(new PrimitiveElement("Horse"));
	comp2->Add(new PrimitiveElement("Dolphin"));
	comp2->Add(new PrimitiveElement("Elephant"));
	root->Add(comp2);

	// Add a primitive element to the second branch
	PrimitiveElement * pe1 = new PrimitiveElement("Cat");
	comp2->Add(pe1);

	// Add a primitive element to the second branch
	comp2->Add(new PrimitiveElement("Dog"));

	// Add a primitive element to the tree
	PrimitiveElement * pe2 = new PrimitiveElement("Sunset at Sea");
	root->Add(pe2);

	// Add a primitive element to first branch
	PrimitiveElement * pe3 = new PrimitiveElement("Orange Trapezoid");
	comp1->Add(pe3);

	// Add a primitive element to the tree
	PrimitiveElement * pe4 = new PrimitiveElement("Golden Horn");
	root->Add(pe4);

	// Remove a primitive element from the secodn branch
	comp2->Remove(pe1);

	// Recursively display nodes
	root->Display(1);

	// Delete the allocated memory
	delete root;

	cin.get();

	return 0;
}

// Output
/*
-+ Paintings
--- Storm
--- Seashore
--- Night in Venice
--- Ninth Wave
---+ Geometric figures
----- Black Circle
----- White Triangle
----- Red Square
----- Blue Line
----- Orange Trapezoid
---+ Animals
----- Horse
----- Dolphin
----- Elephant
----- Dog
--- Sunset at Sea
--- Golden Horn
*/